<a href="https://colab.research.google.com/github/marcarvajalma/AlgorithmsUN2024II/blob/main/Lab3/marcarvajalmaLab3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed bhi
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
ad

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [2]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Overwriting APlusB.cpp


In [3]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

2 2
4

### Test input output files

In [4]:
%%writefile input1.txt
4 8

Overwriting input1.txt


In [5]:
%%writefile output1.txt
12

Overwriting output1.txt


In [6]:
%%writefile input2.txt
0 9

Overwriting input2.txt


In [7]:
%%writefile output2.txt
9

Overwriting output2.txt


In [8]:
!./APlusB < input1.txt

12

In [9]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [10]:
!grep -xvFf output1.txt output1.txt

In [11]:
!grep -xvFf output1.txt output2.txt

9


In [12]:
!grep -xvFf output2.txt output1.txt

12


In [13]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [14]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [15]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Overwriting sum.py


In [16]:
!python sum.py

2 2
4


In [17]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [18]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

$$ T(n) < 6 n^2 = O(n^2)$$

In [19]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Overwriting max_pairwise_product_full.cpp


#### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [20]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

4
1 2 3 4
12


### Test 1: Small numbers

In [21]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Overwriting input0.txt


In [22]:
%%writefile output0.txt
0

Overwriting output0.txt


In [23]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

In [24]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [25]:
!./max_pairwise_product < input1.txt

/bin/bash: line 1: ./max_pairwise_product: No such file or directory


In [26]:
%%writefile output1.txt
90

Overwriting output1.txt


In [27]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [28]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [29]:
%%writefile output2.txt
90

Overwriting output2.txt


In [30]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 to check possible intger overflow

In [31]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Overwriting input3.txt


In [32]:
200000*200000

40000000000

In [33]:
%%writefile output3.txt
40000000000

Overwriting output3.txt


In [34]:
!./max_pairwise_product_full < input3.txt

1345294336


In [35]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [36]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [37]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


### Changing the variables to long long

In [38]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product_long_full.cpp


In [39]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [40]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [41]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [42]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests, 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [43]:
%%writefile input3.txt
6
1 3001 2020 8 22 38

Overwriting input3.txt


In [44]:
%%writefile input4.txt
7
10 437 798 101112 121415 151918 194036

Overwriting input4.txt


In [45]:
%%writefile input5.txt
5
951 753 654 852 379

Overwriting input5.txt


In [46]:
%%writefile input6.txt
8
902 1009 1111 1024 792 564 798 222

Overwriting input6.txt


In [47]:
%%writefile input7.txt
4
741 852 963 684

Overwriting input7.txt


In [48]:
%%writefile input8.txt
5
176 935 378 937 210790

Overwriting input8.txt


In [49]:
%%writefile input9.txt
6
1 3343 2003 9 2222 43

Overwriting input9.txt


In [50]:
%%writefile input10.txt
6
1111 2222 3333 4444 5555 6666

Overwriting input10.txt


In [51]:
!./max_pairwise_product_long_full < input1.txt
!./max_pairwise_product_long_full < input2.txt
!./max_pairwise_product_long_full < input3.txt > output3.txt
!./max_pairwise_product_long_full < input4.txt > output4.txt
!./max_pairwise_product_long_full < input5.txt > output5.txt
!./max_pairwise_product_long_full < input6.txt > output6.txt
!./max_pairwise_product_long_full < input7.txt > output7.txt
!./max_pairwise_product_long_full < input8.txt > output8.txt
!./max_pairwise_product_long_full < input9.txt > output9.txt
!./max_pairwise_product_long_full < input10.txt > output10.txt

90
90


### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and input12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [52]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [53]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [54]:
199999 * 200000

39999800000

In [55]:
%%writefile output11.txt
39999800000

Overwriting output11.txt


In [56]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [57]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [58]:
%%writefile output12.txt
39999800000

Overwriting output12.txt


### Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [59]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

39999800000

real	5m2.467s
user	4m59.334s
sys	0m0.095s


In [60]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

39999800000

real	5m1.718s
user	4m59.090s
sys	0m0.079s


### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [61]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


### Count executed intructions in code

In [62]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Overwriting max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right)  >  6n^2 = O(n^2)$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$


In [63]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

570


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$


In [64]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [65]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [66]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [67]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

600



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [68]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [69]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt

571


Calculate estimated time by instruction and number of instructions by second

For the input11.txt
```
200000
1 2 ..  199999 2000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$

In [70]:
9 + 5*(200000) + 5*(200000)**2 +1

200001000010

In [71]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

200003177091


In [72]:
# Estimated time by instruction

295/200003177091


1.474976569326082e-09

In [73]:
# Estimated number of instructions by second

200003177091/295

677976871.4949152

### Estimate and verify the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [74]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [75]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [76]:
10999 * 11000

120989000

In [77]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m1.500s
user	0m1.428s
sys	0m0.006s


## Python

In [78]:
%%writefile max_pairwise_product_full.py
n = int(input())                        #1
list = input().split()                  #1
list = [int(x) for x in list]           #n

maxProduct = 0                          #1
auxProduct = 0                          #1
for i in range(0,n):                    #1+2n
  for j in range(0,n):                #n+2n^2
    if i!=j:                            #n^2
      auxProduct = list[i]*list[j]      #n^2 -n
    if maxProduct < auxProduct:         #n**2 -n
      maxProduct = auxProduct           #[1, n^2/2 - n/2] upper bound

                                        # --------------
print(maxProduct)                       #5 + 3n + [5*n^2+1, 11*n^2/2  - n/2]  upper bound

Overwriting max_pairwise_product_full.py


PROBAR CON ENTRADAS DE ARCHIVOS DE TEXTO ANTERIORES

In [79]:
!chmod +x max_pairwise_product_full.py

!python max_pairwise_product_full.py < input3.txt > output3.txt
!python max_pairwise_product_full.py < input4.txt > output4.txt
!python max_pairwise_product_full.py < input5.txt > output5.txt
!python max_pairwise_product_full.py < input6.txt > output6.txt
!python max_pairwise_product_full.py < input7.txt > output7.txt
!python max_pairwise_product_full.py < input8.txt > output8.txt
!python max_pairwise_product_full.py < input9.txt > output9.txt
!python max_pairwise_product_full.py < input10.txt > output10.txt

CREAR UN ARCHIVO DE ENTRADA CONSIDERABLEMENTE GRANDE





In [80]:
fileName = "input13Py.txt"

with open(fileName, "w") as file:
    file.write("20000"+ "\n")
    for i in range(1,20001):
        file.write(str(i) + " ")

MEDIR EL TIEMPO

In [81]:
import time

start_time = time.time()
!python max_pairwise_product_full.py < input13Py.txt

end_time = time.time()
execution_time = end_time - start_time
print(f"Tiempo de ejecución en segundos: {execution_time}")
print(f"Tiempo de ejecución en minutos: {execution_time/60}")

399980000
Tiempo de ejecución en segundos: 136.66558170318604
Tiempo de ejecución en minutos: 2.2777596950531005


CONTAR NUMERO DE PASOS

In [82]:
%%writefile max_pairwise_product_full_count.py
contador_instrucciones = 0
cantidad_numeros = int(input())                        #1
contador_instrucciones += 1
lista_numeros = input().split()                        #1
contador_instrucciones += 1
lista_numeros = [int(x) for x in lista_numeros]        #n
contador_instrucciones += cantidad_numeros

producto_maximo = 0                                   #1
contador_instrucciones += 1
producto_auxiliar = 0                                 #1
contador_instrucciones += 1
for i in range(0, cantidad_numeros):                  #1+2n
  contador_instrucciones += 1
  contador_instrucciones += 2
  for j in range(0, cantidad_numeros):                #n + 2n^2
    contador_instrucciones += 2
    contador_instrucciones += 1
    if i != j:                                        #n^2
      producto_auxiliar = lista_numeros[i]*lista_numeros[j]      #n^2 -n
      contador_instrucciones += 1
      contador_instrucciones += 1
    if producto_maximo < producto_auxiliar:           #n^2 -n
      producto_maximo = producto_auxiliar             #n * (n - 1) / 3
      contador_instrucciones += 1

print("Resultado: "+str(producto_maximo))
print("Número de instrucciones: "+str(contador_instrucciones))  #5 + 3n + 3n^2 + 2(n^2 -n)+n * (n - 1) / 3


Overwriting max_pairwise_product_full_count.py


El número de instrucciones ejecutadas por el algoritmo con una lista de tamaño n es:

$$T = 5 + 2n +5( n^{2}) +(n (n - 1) / 3)$$

Depende del tamaño de las listas y el orden de la lista.
Para el siguiente archivo input1.txt

```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 5 + (20) + 5(10)(10) +  90/3= 555$$


In [83]:
!python max_pairwise_product_full_count.py < input1.txt

Resultado: 90
Número de instrucciones: 554


Depende del tamaño de las listas y el orden de la lista.
Entonces para input13Py.txt:

$$T = 5 + (60000) + 3(20000)(20000) + 2(20000)(20000)-20000 + 199990/3= 2000106668$$

In [84]:
!python max_pairwise_product_full_count.py < input13Py.txt

Resultado: 399980000
Número de instrucciones: 2000261148


Calcular el tiempo estimado por instrucción y el número de instrucciones por segundo

In [85]:
%%time
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

600
CPU times: user 9.32 ms, sys: 4.33 ms, total: 13.6 ms
Wall time: 812 ms


In [86]:
%%time
!python max_pairwise_product_full_count.py < input1.txt

Resultado: 90
Número de instrucciones: 554
CPU times: user 5.69 ms, sys: 2.02 ms, total: 7.72 ms
Wall time: 109 ms


In [87]:
tiempo_por_instruccion = 0.0131 / 600
instrucciones_por_segundo = 1 / tiempo_por_instruccion
print(f"Tiempo por instrucción C++: {tiempo_por_instruccion}")
print(f"Instrucciones por segundo C++: {instrucciones_por_segundo}")
tiempo_por_instruccion = 0.00373 / 554
instrucciones_por_segundo = 1 / tiempo_por_instruccion
print(f"Tiempo por instrucción Python: {tiempo_por_instruccion}")
print(f"Instrucciones por segundo Python: {instrucciones_por_segundo}")


Tiempo por instrucción C++: 2.1833333333333333e-05
Instrucciones por segundo C++: 45801.52671755725
Tiempo por instrucción Python: 6.732851985559566e-06
Instrucciones por segundo Python: 148525.4691689008


##Si reemplazamos T en las ecuaciones anteriores, tenemos:

Máximo n que el programa C++ puede resolver en un segundo: 75.8

Máximo n que el programa Python puede resolver en un segundo: 124.51

# For the following solutions required  the same tests and analysis performed for the first solution

## Solutions in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [88]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [89]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!time ./max_pairwise_product< input12.txt

2147483646

real	3m17.444s
user	3m16.006s
sys	0m0.054s


In [90]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Writing max_pairwise_product_long.cpp


El algoritmo ejecuta un número de instrucciones
$$T (n) = 9 + 7n + \left[2(n-1), 2.5(n-1)n\right] \ \  $$

Dependiendo del tamaño y orden de la lista.
Para el archivo input1.txt con una lista de 10 elementos ordenados:

$$T(10) \in [97, 304]  $$


In [91]:
n = 10
print(9 + 7*10+2*(n-1),9 + 7*10+2.5*(n-1)*n)

97 304.0


In [92]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!./max_pairwise_product_long < input1.txt

90


###Se continua en inglés debido a diferencias creativas


[texto del enlace](https://)### Now, in Python

In [93]:
%%writefile max_pairwise_product.py
n = int(input())                        #1
list = input().split()                  #1
list = [int(x) for x in list]           #n

maxProduct = 0                          #1
auxProduct = 0                          #1
for i in range(0,n):                    #1+2n
  for j in range(i+1,n):                #((n**2)-n)/2
    if i!=j:                            #((n**2)-n)/2
      auxProduct = list[i]*list[j]      #((n**2)-n)/2
    if maxProduct < auxProduct:         #((n**2)-n)/2
      maxProduct = auxProduct           #[0,((n**2)-n)/2] upper bound

                                        # --------------
print(maxProduct)                       #5 + n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound

Writing max_pairwise_product.py


Testing with previous txts

In [94]:
!chmod +x max_pairwise_product_py

!python max_pairwise_product.py < input3.txt > output3.txt
!python max_pairwise_product.py < input4.txt > output4.txt
!python max_pairwise_product.py < input5.txt > output5.txt
!python max_pairwise_product.py < input6.txt > output6.txt
!python max_pairwise_product.py < input7.txt > output7.txt
!python max_pairwise_product.py < input8.txt > output8.txt
!python max_pairwise_product.py < input9.txt > output9.txt
!python max_pairwise_product.py < input10.txt > output10.txt

chmod: cannot access 'max_pairwise_product_py': No such file or directory


In [95]:
import time

start_time = time.time()
!python max_pairwise_product.py < inputPy.txt

end_time = time.time()
execution_time = end_time - start_time
print(f"Tiempo de ejecución en segundos: {execution_time}")
print(f"Tiempo de ejecución en minutos: {execution_time/60}")

/bin/bash: line 1: inputPy.txt: No such file or directory
Tiempo de ejecución en segundos: 0.10657167434692383
Tiempo de ejecución en minutos: 0.0017761945724487304


Program to count the number of steps it takes

In [96]:
%%writefile max_pairwise_product_count.py
op = 0
xd =0
n = int(input())                        #1
op+=1
list = input().split()                  #1
op+=1
list = [int(x) for x in list]           #n
op += n

maxProduct = 0                          #1
op+=1
auxProduct = 0                          #1
op+=1
for i in range(0,n):                    #1+2n
  op+=1
  op+=2
  for j in range(i+1,n):                #((n**2)-n)/2
    op+=1
    if i!=j:                            #((n**2)-n)/2
      auxProduct = list[i]*list[j]      #((n**2)-n)/2
      op+=1
      op+=1
    op+=1
    if maxProduct < auxProduct:         #n * (n - 1) / 2
      maxProduct = auxProduct           #n * (n - 1) / 3
      op+=1

                                        # --------------For worst case
print("Result: "+str(maxProduct))
print("Number of instructions: "+str(op))                       #5 + 4n +2( n * (n - 1)) + (n * (n - 1) / 3)

Writing max_pairwise_product_count.py


The number of instructions executed by the algorithm with a list of size $n$ is  

$$T = 5 + 4n +2( n(n - 1)) + (n (n - 1) / 3)$$

It depends on the size of the lists and  the order of the list.

For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 5 + (40) + 2(10)(9) + (90)/3  = 259$$



In [97]:
!python max_pairwise_product_count.py < input1.txt

Result: 90
Number of instructions: 254


For a big input

The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T = 5 + 4n +2( n(n - 1)) + (n (n - 1) / 3)$$

It depends on the size of the lists and  the order of the list.

For the input13Py.txt with 20000 numbers

$$T = 5 + (80000) + 2(20000)(19999) + (199990)/3  = 800106668$$



In [98]:
!python max_pairwise_product_count.py < input13Py.txt

Result: 399980000
Number of instructions: 800261148


###Calculate estimated time by instruction and number of instructions by second

In [99]:
%%time
!g++ max_pairwise_product_long_count.cpp -o max_pairwise_product_long_count
!./max_pairwise_product_long_count < input1.txt

cc1plus: fatal error: max_pairwise_product_long_count.cpp: No such file or directory
compilation terminated.
/bin/bash: line 1: ./max_pairwise_product_long_count: No such file or directory
CPU times: user 10.6 ms, sys: 2.73 ms, total: 13.4 ms
Wall time: 211 ms


In [100]:
%%time
!python max_pairwise_product_count.py < input1.txt

Result: 90
Number of instructions: 254
CPU times: user 6.1 ms, sys: 1.03 ms, total: 7.13 ms
Wall time: 109 ms


In [101]:
time_per_ins = 0.0134 / 231
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction C++: {time_per_ins}")
print(f"Instructions per second C++: {instructions_per_second}")
time_per_ins = 0.0045 / 254
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction Python: {time_per_ins}")
print(f"Instructions per second Python: {instructions_per_second}")

Time per instruction C++: 5.800865800865801e-05
Instructions per second C++: 17238.805970149253
Time per instruction Python: 1.7716535433070864e-05
Instructions per second Python: 56444.44444444445


**If we replace T in the previous equations we have:**

Max n that c++ program can solve in a second: 82.12

Max n that python program can solve in a second: 154.95

### Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

Add in each line a coment with the number of intstrcution that are executed in that line

In [102]:
%%writefile max_pairwise_sort_nlogn.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    cout << "C++" << "\n";
    int n;                                     // 1
    cin >> n;                                  // 1
    vector<int> numbers(n);                    // n

    for (int i = 0; i < n; ++i) {              //1 + 2n
        cin >> numbers[i];                     // 1
    }

    sort(numbers.begin(), numbers.end());      // n*long
    cout <<  "Product: " << numbers[numbers.size() - 1] * numbers[numbers.size() - 2] << "\n"; // 1
    return 0;                                  // 1
}

Writing max_pairwise_sort_nlogn.cpp


In [103]:
%%writefile max_pairwise_sort_nlogn.py
print("Python 3")
# Read input values
n = int(input())              # 1
numbers = list(map(int, input().split()))  # n

# Sort the list
numbers.sort()                # n*log(n)

# Calculate the maximum pairwise product
result = numbers[-1] * numbers[-2]  # 1

# Print the result
print(f"Product: {result}")

Writing max_pairwise_sort_nlogn.py


Test with inputs 1 and 2



In [104]:
!g++ max_pairwise_sort_nlogn.cpp -o max_pairwise_sort_nlogn
!./max_pairwise_sort_nlogn < input1.txt
!./max_pairwise_sort_nlogn < input2.txt
!python max_pairwise_sort_nlogn.py < input1.txt
!python max_pairwise_sort_nlogn.py < input2.txt

C++
Product: 90
C++
Product: 90
Python 3
Product: 90
Python 3
Product: 90


Test and time it with input 3

In [105]:
!g++ max_pairwise_sort_nlogn.cpp -o max_pairwise_sort_nlogn
!./max_pairwise_sort_nlogn < input3.txt

C++
Product: 6062020


20 stress tests

In [106]:
#CREATE 20 random texts
import random

for i in range(1,21):
    fileName = f"stress_input_{i}.txt"

    with open(fileName, "w") as file:
      n = random.randint(3, 40)
      file.write(f"{n}"+ "\n")
      for i in range(1,n):
          file.write(str(random.randint(1, 100)) + " ")


In [107]:
!g++ max_pairwise_sort_nlogn.cpp -o max_pairwise_sort_nlogn
!./max_pairwise_sort_nlogn < stress_input_1.txt
!./max_pairwise_sort_nlogn < stress_input_2.txt
!./max_pairwise_sort_nlogn < stress_input_3.txt
!./max_pairwise_sort_nlogn < stress_input_4.txt
!./max_pairwise_sort_nlogn < stress_input_5.txt
!./max_pairwise_sort_nlogn < stress_input_6.txt
!./max_pairwise_sort_nlogn < stress_input_7.txt
!./max_pairwise_sort_nlogn < stress_input_8.txt
!./max_pairwise_sort_nlogn < stress_input_9.txt
!./max_pairwise_sort_nlogn < stress_input_10.txt
!./max_pairwise_sort_nlogn < stress_input_11.txt
!./max_pairwise_sort_nlogn < stress_input_12.txt
!./max_pairwise_sort_nlogn < stress_input_13.txt
!./max_pairwise_sort_nlogn < stress_input_14.txt
!./max_pairwise_sort_nlogn < stress_input_15.txt
!./max_pairwise_sort_nlogn < stress_input_16.txt
!./max_pairwise_sort_nlogn < stress_input_17.txt
!./max_pairwise_sort_nlogn < stress_input_18.txt
!./max_pairwise_sort_nlogn < stress_input_19.txt
!./max_pairwise_sort_nlogn < stress_input_20.txt


C++
Product: 9024
C++
Product: 7020
C++
Product: 480
C++
Product: 7920
C++
Product: 9312
C++
Product: 9500
C++
Product: 9024
C++
Product: 9310
C++
Product: 5032
C++
Product: 7760
C++
Product: 9604
C++
Product: 7470
C++
Product: 9900
C++
Product: 6640
C++
Product: 9409
C++
Product: 9700
C++
Product: 9212
C++
Product: 4914
C++
Product: 9500
C++
Product: 9700


In [108]:
!python max_pairwise_sort_nlogn.py < stress_input_1.txt
!python max_pairwise_sort_nlogn.py < stress_input_2.txt
!python max_pairwise_sort_nlogn.py < stress_input_3.txt
!python max_pairwise_sort_nlogn.py < stress_input_4.txt
!python max_pairwise_sort_nlogn.py < stress_input_5.txt
!python max_pairwise_sort_nlogn.py < stress_input_6.txt
!python max_pairwise_sort_nlogn.py < stress_input_7.txt
!python max_pairwise_sort_nlogn.py < stress_input_8.txt
!python max_pairwise_sort_nlogn.py < stress_input_9.txt
!python max_pairwise_sort_nlogn.py < stress_input_10.txt
!python max_pairwise_sort_nlogn.py < stress_input_11.txt
!python max_pairwise_sort_nlogn.py < stress_input_12.txt
!python max_pairwise_sort_nlogn.py < stress_input_13.txt
!python max_pairwise_sort_nlogn.py < stress_input_14.txt
!python max_pairwise_sort_nlogn.py < stress_input_15.txt
!python max_pairwise_sort_nlogn.py < stress_input_16.txt
!python max_pairwise_sort_nlogn.py < stress_input_17.txt
!python max_pairwise_sort_nlogn.py < stress_input_18.txt
!python max_pairwise_sort_nlogn.py < stress_input_19.txt
!python max_pairwise_sort_nlogn.py < stress_input_20.txt

Python 3
Product: 9024
Python 3
Product: 7020
Python 3
Product: 480
Python 3
Product: 7920
Python 3
Product: 9312
Python 3
Product: 9500
Python 3
Product: 9024
Python 3
Product: 9310
Python 3
Product: 5032
Python 3
Product: 7760
Python 3
Product: 9604
Python 3
Product: 7470
Python 3
Product: 9900
Python 3
Product: 6640
Python 3
Product: 9409
Python 3
Product: 9700
Python 3
Product: 9212
Python 3
Product: 4914
Python 3
Product: 9500
Python 3
Product: 9700


Count executed instructions in code and verify with inputs 1,2 and 3


In [109]:
%%writefile max_pairwise_sort_nlogn_count.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <bits/stdc++.h>

using namespace std;

int main() {
    cout << "C++" << "\n";

    uint64_t execints = 0;

    int n;                                     // 1
    execints++;
    cin >> n;                                  // 1
    execints++;
    vector<int> numbers(n);                    // n
    execints += n;

    for (int i = 0; i < n; ++i) {              //1 + 2n
        execints += 2;
        cin >> numbers[i];                     // 1
        execints++;
    }

    sort(numbers.begin(), numbers.end());      // n*long

    execints += n * log2(n);

    cout <<  numbers[n - 1] * numbers[n - 2] << "\n"; // 3

    execints += 3;

    cout << "Executed instructions: " << execints << "\n";

    return 0;                                  // 1
}

Writing max_pairwise_sort_nlogn_count.cpp


In [110]:
%%writefile max_pairwise_sort_nlogn_count.py

import math

print("Python 3")

# Read input values

execints = 0
n = int(input())              # 1

execints += 1
numbers = list(map(int, input().split()))  # 4n (n input read + map o(n) + create list + cast to int values)

execints += 4 * n

# Sort the list
numbers.sort()                # n*log(n)

execints += math.log2(n)

# Calculate the maximum pairwise product
result = numbers[-1] * numbers[-2]  # 3

execints += 3

# Print the result
print(f"Executed instructions: {int(execints)}")

Writing max_pairwise_sort_nlogn_count.py


In [111]:
!g++ max_pairwise_sort_nlogn_count.cpp -o max_pairwise_sort_nlogn_count
!./max_pairwise_sort_nlogn_count < input1.txt
!./max_pairwise_sort_nlogn_count < input2.txt
!./max_pairwise_sort_nlogn_count < input3.txt
!python max_pairwise_sort_nlogn_count.py < input1.txt
!python max_pairwise_sort_nlogn_count.py < input2.txt
!python max_pairwise_sort_nlogn_count.py < input3.txt

C++
90
Executed instructions: 78
C++
90
Executed instructions: 78
C++
6062020
Executed instructions: 44
Python 3
Executed instructions: 47
Python 3
Executed instructions: 47
Python 3
Executed instructions: 30


Calculate estimated time by instruction and number of instructions by second

In [112]:
%%time
!g++ max_pairwise_sort_nlogn_count.cpp -o max_pairwise_sort_nlogn_count
!./max_pairwise_sort_nlogn_count < input1.txt

C++
90
Executed instructions: 78
CPU times: user 18.4 ms, sys: 5.15 ms, total: 23.5 ms
Wall time: 2.12 s


In [113]:
%%time
!python max_pairwise_sort_nlogn_count.py < input1.txt

Python 3
Executed instructions: 47
CPU times: user 4.8 ms, sys: 2.05 ms, total: 6.85 ms
Wall time: 106 ms


In [114]:
time_per_ins = 0.025 / 78
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction C++: {time_per_ins}")
print(f"Instructions per second C++: {instructions_per_second}")
time_per_ins = 0.00701 / 47
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction Python: {time_per_ins}")
print(f"Instructions per second Python: {instructions_per_second}")

Time per instruction C++: 0.0003205128205128205
Instructions per second C++: 3120.0
Time per instruction Python: 0.00014914893617021276
Instructions per second Python: 6704.707560627675


### Efficient solution largest and second-largest in an list, array or vector in C++ and Python
 $$2.5 n = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [115]:
%%writefile max_pairwise_sort_efficient.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits.h>

using namespace std;

int main() {
    cout << "--- C++ ---" << "\n";
    int n;                                     // 1
    cin >> n;                                  // 1
    int maxNumber = INT_MIN;                   // 1
    int secondMaxNumber = INT_MIN;             // 1

    vector<int> numbers(n);                    // n

    for (int i = 0; i < n; ++i) {              // 1 + 2n
        cin >> numbers[i];                     // n
    }


    for (int i = 0; i < n; ++i) {              //1 + 2n
        int aux = numbers[i];                           // n

        if (aux > maxNumber){                  // n
          secondMaxNumber = maxNumber;         // [0, n] in the worst case it would be n times
          maxNumber = aux;                     // [0, n] in the worst case it would be n times

        } else if (aux > secondMaxNumber) {    // n
          secondMaxNumber = aux;               // [0, n] in the worst case it would be n times
        }

    }

    cout <<  "Product: " << maxNumber * secondMaxNumber << "\n"; // 1
    return 0;                                  // 1
}

Writing max_pairwise_sort_efficient.cpp


In [116]:
%%writefile max_pairwise_sort_efficient.py
print("--- Python ---")
n = int(input())
numbers = list(map(int, input().split()))

max_number = float("-inf")                   # 1
second_max_number = float("-inf")            # 1

for num in numbers:                          # 1 + 2n
    if num > max_number:                     # n
        second_max_number = max_number       # [0, n] in the worst case it would be n times
        max_number = num                     # [0, n] in the worst case it would be n times
    elif num > second_max_number:            # n
        second_max_number = num              # [0, n] in the worst case it would be n times

product = max_number * second_max_number     # 1
print(f"Product: {product}")                 # 1

Writing max_pairwise_sort_efficient.py


Test with inputs 1 and 2



In [117]:
!g++ max_pairwise_sort_efficient.cpp -o max_pairwise_sort_efficient
!./max_pairwise_sort_efficient < input1.txt
!./max_pairwise_sort_efficient < input2.txt
! python max_pairwise_sort_efficient.py < input1.txt
! python max_pairwise_sort_efficient.py < input2.txt

--- C++ ---
Product: 90
--- C++ ---
Product: 90
--- Python ---
Product: 90
--- Python ---
Product: 90


Test and time it with input 3

In [118]:
!g++ max_pairwise_sort_efficient.cpp -o max_pairwise_sort_efficient
!./max_pairwise_sort_efficient < input3.txt
! python max_pairwise_sort_efficient.py < input3.txt

--- C++ ---
Product: 6062020
--- Python ---
Product: 6062020


***20*** stress tests

In [119]:
#CREATE 20 random texts
import random

for i in range(1,21):
    fileName = f"stress_input_{i}.txt"

    with open(fileName, "w") as file:
      n = random.randint(3, 40)
      file.write(f"{n}"+ "\n")
      for i in range(1,n):
          file.write(str(random.randint(1, 100)) + " ")

In [120]:
!g++ max_pairwise_sort_efficient.cpp -o max_pairwise_sort_efficient
!./max_pairwise_sort_efficient < stress_input_1.txt
!./max_pairwise_sort_efficient < stress_input_2.txt
!./max_pairwise_sort_efficient < stress_input_3.txt
!./max_pairwise_sort_efficient < stress_input_4.txt
!./max_pairwise_sort_efficient < stress_input_5.txt
!./max_pairwise_sort_efficient < stress_input_6.txt
!./max_pairwise_sort_efficient < stress_input_7.txt
!./max_pairwise_sort_efficient < stress_input_8.txt
!./max_pairwise_sort_efficient < stress_input_9.txt
!./max_pairwise_sort_efficient < stress_input_10.txt
!./max_pairwise_sort_efficient < stress_input_11.txt
!./max_pairwise_sort_efficient < stress_input_12.txt
!./max_pairwise_sort_efficient < stress_input_13.txt
!./max_pairwise_sort_efficient < stress_input_14.txt
!./max_pairwise_sort_efficient < stress_input_15.txt
!./max_pairwise_sort_efficient < stress_input_16.txt
!./max_pairwise_sort_efficient < stress_input_17.txt
!./max_pairwise_sort_efficient < stress_input_18.txt
!./max_pairwise_sort_efficient < stress_input_19.txt
!./max_pairwise_sort_efficient < stress_input_20.txt

--- C++ ---
Product: 3816
--- C++ ---
Product: 8544
--- C++ ---
Product: 9900
--- C++ ---
Product: 9500
--- C++ ---
Product: 9021
--- C++ ---
Product: 2849
--- C++ ---
Product: 8835
--- C++ ---
Product: 8184
--- C++ ---
Product: 9600
--- C++ ---
Product: 9216
--- C++ ---
Product: 8366
--- C++ ---
Product: 8188
--- C++ ---
Product: 9900
--- C++ ---
Product: 465
--- C++ ---
Product: 8051
--- C++ ---
Product: 8712
--- C++ ---
Product: 8265
--- C++ ---
Product: 8051
--- C++ ---
Product: 1702
--- C++ ---
Product: 8736


In [121]:
! python max_pairwise_sort_efficient.py < stress_input_1.txt
! python max_pairwise_sort_efficient.py < stress_input_2.txt
! python max_pairwise_sort_efficient.py < stress_input_3.txt
! python max_pairwise_sort_efficient.py < stress_input_4.txt
! python max_pairwise_sort_efficient.py < stress_input_5.txt
! python max_pairwise_sort_efficient.py < stress_input_6.txt
! python max_pairwise_sort_efficient.py < stress_input_7.txt
! python max_pairwise_sort_efficient.py < stress_input_8.txt
! python max_pairwise_sort_efficient.py < stress_input_9.txt
! python max_pairwise_sort_efficient.py < stress_input_10.txt
! python max_pairwise_sort_efficient.py < stress_input_11.txt
! python max_pairwise_sort_efficient.py < stress_input_12.txt
! python max_pairwise_sort_efficient.py < stress_input_13.txt
! python max_pairwise_sort_efficient.py < stress_input_14.txt
! python max_pairwise_sort_efficient.py < stress_input_15.txt
! python max_pairwise_sort_efficient.py < stress_input_16.txt
! python max_pairwise_sort_efficient.py < stress_input_17.txt
! python max_pairwise_sort_efficient.py < stress_input_18.txt
! python max_pairwise_sort_efficient.py < stress_input_19.txt
! python max_pairwise_sort_efficient.py < stress_input_20.txt

--- Python ---
Product: 3816
--- Python ---
Product: 8544
--- Python ---
Product: 9900
--- Python ---
Product: 9500
--- Python ---
Product: 9021
--- Python ---
Product: 2849
--- Python ---
Product: 8835
--- Python ---
Product: 8184
--- Python ---
Product: 9600
--- Python ---
Product: 9216
--- Python ---
Product: 8366
--- Python ---
Product: 8188
--- Python ---
Product: 9900
--- Python ---
Product: 465
--- Python ---
Product: 8051
--- Python ---
Product: 8712
--- Python ---
Product: 8265
--- Python ---
Product: 8051
--- Python ---
Product: 1702
--- Python ---
Product: 8736


Count executed instructions in code and verify with inputs 1,2 and 3


In [122]:
%%writefile max_pairwise_sort_efficient_count.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits.h>

using namespace std;

int main() {
    cout << "--- C++ ---" << "\n";
    int exexins = 0;
    int n;                                     // 1
    exexins++;
    cin >> n;                                  // 1
    exexins++;
    int maxNumber = INT_MIN;                   // 1
    exexins++;
    int secondMaxNumber = INT_MIN;             // 1
    exexins++;

    vector<int> numbers(n);                    // n

    exexins += n;

    for (int i = 0; i < n; ++i) {              // 1 + 2n
        cin >> numbers[i];                     // n
        exexins++;
    }


    exexins++;
    for (int i = 0; i < n; ++i) {              //1 + 2n
        exexins += 2;

        int aux = numbers[i];                           // n
        exexins++;


        exexins++; // if
        exexins++; // else if
        if (aux > maxNumber){                  // n
          secondMaxNumber = maxNumber;         // [0, n] in the worst case it would be n times
          exexins++;
          maxNumber = aux;                     // [0, n] in the worst case it would be n times
          exexins++;
        } else if (aux > secondMaxNumber) {    // n
          secondMaxNumber = aux;               // [0, n] in the worst case it would be n times
          exexins++;
        }

    }

    cout <<  "Executed instructions: " << exexins << " n: "<< n << " Product: " << maxNumber * secondMaxNumber << "\n"; // 1
    return 0;                                  // 1
}

Writing max_pairwise_sort_efficient_count.cpp


In [123]:
!g++ max_pairwise_sort_efficient_count.cpp -o max_pairwise_sort_efficient_count
!./max_pairwise_sort_efficient_count < input1.txt
!./max_pairwise_sort_efficient_count < input2.txt
!./max_pairwise_sort_efficient_count < input3.txt

--- C++ ---
Executed instructions: 95 n: 10 Product: 90
--- C++ ---
Executed instructions: 78 n: 10 Product: 90
--- C++ ---
Executed instructions: 52 n: 6 Product: 6062020


In [124]:
%%writefile max_pairwise_sort_efficient_count.py
print("--- Python ---")
exexins = 0
n = int(input())                             # 1
exexins += 1
numbers = list(map(int, input().split()))    # 4n (n input read + map o(n) + create list + cast to int values)
exexins = 4 * n

max_number = float("-inf")                   # 1
exexins += 1
second_max_number = float("-inf")            # 1
exexins += 1

exexins += 1
for num in numbers:                          # 1 + 2n
    exexins += 2

    exexins += 1 # if
    exexins += 1 # elif
    if num > max_number:                     # n
        second_max_number = max_number       # [0, n] in the worst case it would be n times
        exexins += 1
        max_number = num                     # [0, n] in the worst case it would be n times
        exexins += 1
    elif num > second_max_number:            # n
        second_max_number = num              # [0, n] in the worst case it would be n times
        exexins += 1

product = max_number * second_max_number     # 1
exexins += 1
print(f"Executed instructions: {exexins} n: {n} Product: {product}")                 # 1


Writing max_pairwise_sort_efficient_count.py


In [125]:
!python max_pairwise_sort_efficient_count.py < input1.txt
!python max_pairwise_sort_efficient_count.py < input2.txt
!python max_pairwise_sort_efficient_count.py < input3.txt

--- Python ---
Executed instructions: 104 n: 10 Product: 90
--- Python ---
Executed instructions: 87 n: 10 Product: 90
--- Python ---
Executed instructions: 57 n: 6 Product: 6062020


Calculate estimated time by instruction and number of instructions by second

In [126]:
%%time
!g++ max_pairwise_sort_efficient_count.cpp -o max_pairwise_sort_efficient_count
!./max_pairwise_sort_efficient_count < input1.txt

--- C++ ---
Executed instructions: 95 n: 10 Product: 90
CPU times: user 14.1 ms, sys: 2.8 ms, total: 16.9 ms
Wall time: 817 ms


In [127]:
%%time
!python max_pairwise_sort_efficient_count.py < input1.txt

--- Python ---
Executed instructions: 104 n: 10 Product: 90
CPU times: user 3.83 ms, sys: 3.82 ms, total: 7.66 ms
Wall time: 107 ms


In [128]:
time_per_ins = 0.0115 / 85
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction C++: {time_per_ins}")
print(f"Instructions per second C++: {instructions_per_second}")
time_per_ins = 0.00846 / 104
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction Python: {time_per_ins}")
print(f"Instructions per second Python: {instructions_per_second}")

Time per instruction C++: 0.0001352941176470588
Instructions per second C++: 7391.304347826088
Time per instruction Python: 8.134615384615385e-05
Instructions per second Python: 12293.144208037824


### Super efficient solution IN C++ AND PYTHON (both)

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python
 $$2 n = O(n)$$

In [129]:
%%writefile max_pairwise_sort_super_efficient.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits.h>

using namespace std;

int main() {
    cout << "--- C++ ---" << "\n";
    int n;                                     // 1
    cin >> n;                                  // 1
    int maxNumber = INT_MIN;                   // 1
    int secondMaxNumber = INT_MIN;             // 1


    for (int i = 0; i < n; ++i) {              //1 + 2n
        int aux = 0;                           // n
        cin >> aux;                            // n

        if (aux > maxNumber){                  // n
          secondMaxNumber = maxNumber;         // [0, n] in the worst case it would be n times
          maxNumber = aux;                     // [0, n] in the worst case it would be n times

        } else if (aux > secondMaxNumber) {    // n
          secondMaxNumber = aux;               // [0, n] in the worst case it would be n times
        }

    }

    cout <<  "Product: " << maxNumber * secondMaxNumber << "\n"; // 1
    return 0;                                  // 1
}

Writing max_pairwise_sort_super_efficient.cpp


In [130]:
%%writefile max_pairwise_sort_super_efficient.py
print("--- Python ---")
n = int(input())                            # 1
max_number = float("-inf")                  # 1
second_max_number = float("-inf")           # 1

for num in input().split():                 # 3 + 2n

    num = int(num)                          # n

    if num > max_number:                    # n
        second_max_number = max_number      # [0, n] in the worst case it would be n times
        max_number = num                    # [0, n] in the worst case it would be n times
    elif num > second_max_number:           # n
        second_max_number = num             # [0, n] in the worst case it would be n times

print(f"Product: {max_number * second_max_number}")

Writing max_pairwise_sort_super_efficient.py


Test with inputs 1 and 2




In [131]:
!g++ max_pairwise_sort_super_efficient.cpp -o max_pairwise_sort_super_efficient
!./max_pairwise_sort_super_efficient < input1.txt
!./max_pairwise_sort_super_efficient < input2.txt
!python max_pairwise_sort_super_efficient.py < input1.txt
!python max_pairwise_sort_super_efficient.py < input2.txt

--- C++ ---
Product: 90
--- C++ ---
Product: 90
--- Python ---
Product: 90
--- Python ---
Product: 90


Test and time it with input 3

In [132]:
!g++ max_pairwise_sort_super_efficient.cpp -o max_pairwise_sort_super_efficient
!./max_pairwise_sort_super_efficient < input3.txt
!python max_pairwise_sort_super_efficient.py < input3.txt

--- C++ ---
Product: 6062020
--- Python ---
Product: 6062020


20 stress tests

In [133]:
#CREATE 20 random texts
import random

for i in range(1,21):
    fileName = f"stress_input_{i}.txt"

    with open(fileName, "w") as file:
      n = random.randint(3, 40)
      file.write(f"{n}"+ "\n")
      for i in range(1,n):
          file.write(str(random.randint(1, 100)) + " ")

In [134]:
!g++ max_pairwise_sort_super_efficient.cpp -o max_pairwise_sort_super_efficient
!./max_pairwise_sort_super_efficient < stress_input_1.txt
!./max_pairwise_sort_super_efficient < stress_input_2.txt
!./max_pairwise_sort_super_efficient < stress_input_3.txt
!./max_pairwise_sort_super_efficient < stress_input_4.txt
!./max_pairwise_sort_super_efficient < stress_input_5.txt
!./max_pairwise_sort_super_efficient < stress_input_6.txt
!./max_pairwise_sort_super_efficient < stress_input_7.txt
!./max_pairwise_sort_super_efficient < stress_input_8.txt
!./max_pairwise_sort_super_efficient < stress_input_9.txt
!./max_pairwise_sort_super_efficient < stress_input_10.txt
!./max_pairwise_sort_super_efficient < stress_input_11.txt
!./max_pairwise_sort_super_efficient < stress_input_12.txt
!./max_pairwise_sort_super_efficient < stress_input_13.txt
!./max_pairwise_sort_super_efficient < stress_input_14.txt
!./max_pairwise_sort_super_efficient < stress_input_15.txt
!./max_pairwise_sort_super_efficient < stress_input_16.txt
!./max_pairwise_sort_super_efficient < stress_input_17.txt
!./max_pairwise_sort_super_efficient < stress_input_18.txt
!./max_pairwise_sort_super_efficient < stress_input_19.txt
!./max_pairwise_sort_super_efficient < stress_input_20.txt

--- C++ ---
Product: 1020
--- C++ ---
Product: 7225
--- C++ ---
Product: 8900
--- C++ ---
Product: 9800
--- C++ ---
Product: 8036
--- C++ ---
Product: 9700
--- C++ ---
Product: 9108
--- C++ ---
Product: 9800
--- C++ ---
Product: 9215
--- C++ ---
Product: 8836
--- C++ ---
Product: 9800
--- C++ ---
Product: 9312
--- C++ ---
Product: 9408
--- C++ ---
Product: 9310
--- C++ ---
Product: 2340
--- C++ ---
Product: 8924
--- C++ ---
Product: 8500
--- C++ ---
Product: 9405
--- C++ ---
Product: 8742
--- C++ ---
Product: 9024


In [135]:
!python max_pairwise_sort_super_efficient.py < stress_input_1.txt
!python max_pairwise_sort_super_efficient.py < stress_input_2.txt
!python max_pairwise_sort_super_efficient.py < stress_input_3.txt
!python max_pairwise_sort_super_efficient.py < stress_input_4.txt
!python max_pairwise_sort_super_efficient.py < stress_input_5.txt
!python max_pairwise_sort_super_efficient.py < stress_input_6.txt
!python max_pairwise_sort_super_efficient.py < stress_input_7.txt
!python max_pairwise_sort_super_efficient.py < stress_input_8.txt
!python max_pairwise_sort_super_efficient.py < stress_input_9.txt
!python max_pairwise_sort_super_efficient.py < stress_input_10.txt
!python max_pairwise_sort_super_efficient.py < stress_input_11.txt
!python max_pairwise_sort_super_efficient.py < stress_input_12.txt
!python max_pairwise_sort_super_efficient.py < stress_input_13.txt
!python max_pairwise_sort_super_efficient.py < stress_input_14.txt
!python max_pairwise_sort_super_efficient.py < stress_input_15.txt
!python max_pairwise_sort_super_efficient.py < stress_input_16.txt
!python max_pairwise_sort_super_efficient.py < stress_input_17.txt
!python max_pairwise_sort_super_efficient.py < stress_input_18.txt
!python max_pairwise_sort_super_efficient.py < stress_input_19.txt
!python max_pairwise_sort_super_efficient.py < stress_input_20.txt

--- Python ---
Product: 1020
--- Python ---
Product: 7225
--- Python ---
Product: 8900
--- Python ---
Product: 9800
--- Python ---
Product: 8036
--- Python ---
Product: 9700
--- Python ---
Product: 9108
--- Python ---
Product: 9800
--- Python ---
Product: 9215
--- Python ---
Product: 8836
--- Python ---
Product: 9800
--- Python ---
Product: 9312
--- Python ---
Product: 9408
--- Python ---
Product: 9310
--- Python ---
Product: 2340
--- Python ---
Product: 8924
--- Python ---
Product: 8500
--- Python ---
Product: 9405
--- Python ---
Product: 8742
--- Python ---
Product: 9024


Count executed instructions in code and verify with inputs 1,2 and 3

In [136]:
%%writefile max_pairwise_sort_super_efficient_count.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits.h>

using namespace std;

int main() {
    cout << "--- C++ ---" << "\n";
    int exexins = 0;
    int n;                                     // 1
    exexins++;
    cin >> n;                                  // 1
    exexins++;
    int maxNumber = INT_MIN;                   // 1
    exexins++;
    int secondMaxNumber = INT_MIN;             // 1
    exexins++;


    exexins++;
    for (int i = 0; i < n; ++i) {              //1 + 2n
        exexins += 2;
        int aux = 0;                           // n
        exexins++;

        cin>> aux;                            // n
        exexins++;


        exexins++; // if
        exexins++; // else if
        if (aux > maxNumber){                  // n
          secondMaxNumber = maxNumber;         // [0, n] in the worst case it would be n times
          exexins++;
          maxNumber = aux;                     // [0, n] in the worst case it would be n times
          exexins++;
        } else if (aux > secondMaxNumber) {    // n
          secondMaxNumber = aux;               // [0, n] in the worst case it would be n times
          exexins++;
        }

    }

    cout <<  "Executed instructions: " << exexins << " n: "<< n << " Product: " << maxNumber * secondMaxNumber << "\n"; // 1
    return 0;                                  // 1
}

Writing max_pairwise_sort_super_efficient_count.cpp


In [137]:
%%writefile max_pairwise_sort_super_efficient_count.py
print("--- Python ---")
exexins = 0
n = int(input())                            # 1
exexins += 1
max_number = float("-inf")                  # 1
exexins += 1
second_max_number = float("-inf")           # 1
exexins += 1

exexins += 3
for num in input().split():                 # 3 + 2n
    exexins += 2
    num = int(num)                          # n
    exexins += 1

    exexins += 1 # if
    exexins += 1 # elif
    if num > max_number:                    # n
        second_max_number = max_number      # [0, n] in the worst case it would be n times
        exexins += 1
        max_number = num                    # [0, n] in the worst case it would be n times
        exexins += 1
    elif num > second_max_number:           # n
        second_max_number = num             # [0, n] in the worst case it would be n times
        exexins += 1

print(f"Executed instructions: {exexins} n: {n} Product: {max_number * second_max_number}")

Writing max_pairwise_sort_super_efficient_count.py


In [138]:
!g++ max_pairwise_sort_super_efficient_count.cpp -o max_pairwise_sort_super_efficient_count
!./max_pairwise_sort_super_efficient_count < input1.txt
!./max_pairwise_sort_super_efficient_count < input2.txt
!./max_pairwise_sort_super_efficient_count < input3.txt
!python max_pairwise_sort_super_efficient_count.py < input1.txt
!python max_pairwise_sort_super_efficient_count.py < input2.txt
!python max_pairwise_sort_super_efficient_count.py < input3.txt

--- C++ ---
Executed instructions: 85 n: 10 Product: 90
--- C++ ---
Executed instructions: 68 n: 10 Product: 90
--- C++ ---
Executed instructions: 46 n: 6 Product: 6062020
--- Python ---
Executed instructions: 76 n: 10 Product: 90
--- Python ---
Executed instructions: 59 n: 10 Product: 90
--- Python ---
Executed instructions: 41 n: 6 Product: 6062020


Calculate estimated time by instruction and number of instructions by second

In [139]:
%%time
!g++ max_pairwise_sort_super_efficient_count.cpp -o max_pairwise_sort_super_efficient_count
!./max_pairwise_sort_super_efficient_count < input1.txt

--- C++ ---
Executed instructions: 85 n: 10 Product: 90
CPU times: user 11.8 ms, sys: 3.33 ms, total: 15.2 ms
Wall time: 815 ms


In [140]:
%%time
!python max_pairwise_sort_super_efficient_count.py < input1.txt

--- Python ---
Executed instructions: 76 n: 10 Product: 90
CPU times: user 8.96 ms, sys: 0 ns, total: 8.96 ms
Wall time: 106 ms


In [141]:
time_per_ins = 0.011 / 85
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction C++: {time_per_ins}")
print(f"Instructions per second C++: {instructions_per_second}")
time_per_ins = 0.00361 / 76
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction Python: {time_per_ins}")
print(f"Instructions per second Python: {instructions_per_second}")

Time per instruction C++: 0.00012941176470588234
Instructions per second C++: 7727.272727272728
Time per instruction Python: 4.7499999999999996e-05
Instructions per second Python: 21052.63157894737


# Perform 100 randomized automated tests in C++ and Python between the super efficient solution  (𝑇(𝑛)<10𝑛=𝑂(𝑛))  against the sorting solution  (𝑇(𝑛)=𝑂(𝑛𝑙𝑛𝑔)) .

In [142]:
#CREATE 100 random texts
import random

for i in range(1,101):
    fileName = f"stress_input_{i}.txt"

    with open(fileName, "w") as file:
      n = random.randint(3, 40)
      file.write(f"{n}"+ "\n")
      for i in range(1,n):
          file.write(str(random.randint(1, 100)) + " ")

# C++

## Super efficient solution (𝑇(𝑛)<10𝑛=𝑂(𝑛))

In [143]:
%%writefile max_pairwise_sort_super_efficient.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits.h>

using namespace std;

int main() {
    int n;                                     // 1
    cin >> n;                                  // 1
    int maxNumber = INT_MIN;                   // 1
    int secondMaxNumber = INT_MIN;             // 1


    for (int i = 0; i < n; ++i) {              //1 + 2n
        int aux = 0;                           // n
        cin >> aux;                            // n

        if (aux > maxNumber){                  // n
          secondMaxNumber = maxNumber;         // [0, n] in the worst case it would be n times
          maxNumber = aux;                     // [0, n] in the worst case it would be n times

        } else if (aux > secondMaxNumber) {    // n
          secondMaxNumber = aux;               // [0, n] in the worst case it would be n times
        }

    }

    cout <<  "n: " << n << " Product: " << maxNumber * secondMaxNumber << "\n"; // 1
    return 0;                                  // 1
}

Overwriting max_pairwise_sort_super_efficient.cpp


In [144]:
%%time
!g++ max_pairwise_sort_super_efficient.cpp -o max_pairwise_sort_super_efficient
!./max_pairwise_sort_super_efficient < stress_input_1.txt
!./max_pairwise_sort_super_efficient < stress_input_2.txt
!./max_pairwise_sort_super_efficient < stress_input_3.txt
!./max_pairwise_sort_super_efficient < stress_input_4.txt
!./max_pairwise_sort_super_efficient < stress_input_5.txt
!./max_pairwise_sort_super_efficient < stress_input_6.txt
!./max_pairwise_sort_super_efficient < stress_input_7.txt
!./max_pairwise_sort_super_efficient < stress_input_8.txt
!./max_pairwise_sort_super_efficient < stress_input_9.txt
!./max_pairwise_sort_super_efficient < stress_input_10.txt

!./max_pairwise_sort_super_efficient < stress_input_11.txt
!./max_pairwise_sort_super_efficient < stress_input_12.txt
!./max_pairwise_sort_super_efficient < stress_input_13.txt
!./max_pairwise_sort_super_efficient < stress_input_14.txt
!./max_pairwise_sort_super_efficient < stress_input_15.txt
!./max_pairwise_sort_super_efficient < stress_input_16.txt
!./max_pairwise_sort_super_efficient < stress_input_17.txt
!./max_pairwise_sort_super_efficient < stress_input_18.txt
!./max_pairwise_sort_super_efficient < stress_input_19.txt
!./max_pairwise_sort_super_efficient < stress_input_20.txt

!./max_pairwise_sort_super_efficient < stress_input_21.txt
!./max_pairwise_sort_super_efficient < stress_input_22.txt
!./max_pairwise_sort_super_efficient < stress_input_23.txt
!./max_pairwise_sort_super_efficient < stress_input_24.txt
!./max_pairwise_sort_super_efficient < stress_input_25.txt
!./max_pairwise_sort_super_efficient < stress_input_26.txt
!./max_pairwise_sort_super_efficient < stress_input_27.txt
!./max_pairwise_sort_super_efficient < stress_input_28.txt
!./max_pairwise_sort_super_efficient < stress_input_29.txt
!./max_pairwise_sort_super_efficient < stress_input_30.txt

!./max_pairwise_sort_super_efficient < stress_input_31.txt
!./max_pairwise_sort_super_efficient < stress_input_32.txt
!./max_pairwise_sort_super_efficient < stress_input_33.txt
!./max_pairwise_sort_super_efficient < stress_input_34.txt
!./max_pairwise_sort_super_efficient < stress_input_35.txt
!./max_pairwise_sort_super_efficient < stress_input_36.txt
!./max_pairwise_sort_super_efficient < stress_input_37.txt
!./max_pairwise_sort_super_efficient < stress_input_38.txt
!./max_pairwise_sort_super_efficient < stress_input_39.txt
!./max_pairwise_sort_super_efficient < stress_input_40.txt

!./max_pairwise_sort_super_efficient < stress_input_41.txt
!./max_pairwise_sort_super_efficient < stress_input_42.txt
!./max_pairwise_sort_super_efficient < stress_input_43.txt
!./max_pairwise_sort_super_efficient < stress_input_44.txt
!./max_pairwise_sort_super_efficient < stress_input_45.txt
!./max_pairwise_sort_super_efficient < stress_input_46.txt
!./max_pairwise_sort_super_efficient < stress_input_47.txt
!./max_pairwise_sort_super_efficient < stress_input_48.txt
!./max_pairwise_sort_super_efficient < stress_input_49.txt
!./max_pairwise_sort_super_efficient < stress_input_50.txt

!./max_pairwise_sort_super_efficient < stress_input_51.txt
!./max_pairwise_sort_super_efficient < stress_input_52.txt
!./max_pairwise_sort_super_efficient < stress_input_53.txt
!./max_pairwise_sort_super_efficient < stress_input_54.txt
!./max_pairwise_sort_super_efficient < stress_input_55.txt
!./max_pairwise_sort_super_efficient < stress_input_56.txt
!./max_pairwise_sort_super_efficient < stress_input_57.txt
!./max_pairwise_sort_super_efficient < stress_input_58.txt
!./max_pairwise_sort_super_efficient < stress_input_59.txt
!./max_pairwise_sort_super_efficient < stress_input_60.txt

!./max_pairwise_sort_super_efficient < stress_input_61.txt
!./max_pairwise_sort_super_efficient < stress_input_62.txt
!./max_pairwise_sort_super_efficient < stress_input_63.txt
!./max_pairwise_sort_super_efficient < stress_input_64.txt
!./max_pairwise_sort_super_efficient < stress_input_65.txt
!./max_pairwise_sort_super_efficient < stress_input_66.txt
!./max_pairwise_sort_super_efficient < stress_input_67.txt
!./max_pairwise_sort_super_efficient < stress_input_68.txt
!./max_pairwise_sort_super_efficient < stress_input_69.txt
!./max_pairwise_sort_super_efficient < stress_input_70.txt

!./max_pairwise_sort_super_efficient < stress_input_71.txt
!./max_pairwise_sort_super_efficient < stress_input_72.txt
!./max_pairwise_sort_super_efficient < stress_input_73.txt
!./max_pairwise_sort_super_efficient < stress_input_74.txt
!./max_pairwise_sort_super_efficient < stress_input_75.txt
!./max_pairwise_sort_super_efficient < stress_input_76.txt
!./max_pairwise_sort_super_efficient < stress_input_77.txt
!./max_pairwise_sort_super_efficient < stress_input_78.txt
!./max_pairwise_sort_super_efficient < stress_input_79.txt
!./max_pairwise_sort_super_efficient < stress_input_80.txt

!./max_pairwise_sort_super_efficient < stress_input_81.txt
!./max_pairwise_sort_super_efficient < stress_input_82.txt
!./max_pairwise_sort_super_efficient < stress_input_83.txt
!./max_pairwise_sort_super_efficient < stress_input_84.txt
!./max_pairwise_sort_super_efficient < stress_input_85.txt
!./max_pairwise_sort_super_efficient < stress_input_86.txt
!./max_pairwise_sort_super_efficient < stress_input_87.txt
!./max_pairwise_sort_super_efficient < stress_input_88.txt
!./max_pairwise_sort_super_efficient < stress_input_89.txt
!./max_pairwise_sort_super_efficient < stress_input_90.txt

!./max_pairwise_sort_super_efficient < stress_input_91.txt
!./max_pairwise_sort_super_efficient < stress_input_92.txt
!./max_pairwise_sort_super_efficient < stress_input_93.txt
!./max_pairwise_sort_super_efficient < stress_input_94.txt
!./max_pairwise_sort_super_efficient < stress_input_95.txt
!./max_pairwise_sort_super_efficient < stress_input_96.txt
!./max_pairwise_sort_super_efficient < stress_input_97.txt
!./max_pairwise_sort_super_efficient < stress_input_98.txt
!./max_pairwise_sort_super_efficient < stress_input_99.txt
!./max_pairwise_sort_super_efficient < stress_input_100.txt


n: 16 Product: 9025
n: 38 Product: 9024
n: 10 Product: 7728
n: 40 Product: 9300
n: 18 Product: 8366
n: 27 Product: 7920
n: 5 Product: 5160
n: 32 Product: 9312
n: 39 Product: 8835
n: 21 Product: 8740
n: 12 Product: 9215
n: 4 Product: 7990
n: 25 Product: 9800
n: 16 Product: 8190
n: 33 Product: 9409
n: 28 Product: 9504
n: 40 Product: 9603
n: 21 Product: 7031
n: 29 Product: 8099
n: 13 Product: 8051
n: 20 Product: 5508
n: 4 Product: 735
n: 22 Product: 9801
n: 29 Product: 9900
n: 36 Product: 9603
n: 4 Product: 1566
n: 36 Product: 9408
n: 25 Product: 9306
n: 32 Product: 8918
n: 9 Product: 7857
n: 17 Product: 8556
n: 34 Product: 10000
n: 19 Product: 7921
n: 3 Product: 2508
n: 29 Product: 8184
n: 8 Product: 8245
n: 14 Product: 8148
n: 17 Product: 7743
n: 10 Product: 8463
n: 12 Product: 7568
n: 10 Product: 4050
n: 38 Product: 9408
n: 12 Product: 6016
n: 13 Product: 8272
n: 8 Product: 7920
n: 28 Product: 8930
n: 32 Product: 9603
n: 26 Product: 8256
n: 30 Product: 8134
n: 9 Product: 5832
n: 13 Pro

## Sortin solution (𝑇(𝑛)=𝑂(𝑛𝑙𝑛𝑔))

In [145]:
%%writefile max_pairwise_sort_nlogn.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    int n;                                     // 1
    cin >> n;                                  // 1
    vector<int> numbers(n);                    // n

    for (int i = 0; i < n; ++i) {              //1 + 2n
        cin >> numbers[i];                     // 1
    }

    sort(numbers.begin(), numbers.end());      // n*long
    cout <<  "n: " << n << " Product: " << numbers[numbers.size() - 1] * numbers[numbers.size() - 2] << "\n"; // 1
    return 0;                                  // 1
}

Overwriting max_pairwise_sort_nlogn.cpp


In [146]:
%%time
!g++ max_pairwise_sort_nlogn.cpp -o max_pairwise_sort_nlogn
!./max_pairwise_sort_nlogn < stress_input_1.txt
!./max_pairwise_sort_nlogn < stress_input_2.txt
!./max_pairwise_sort_nlogn < stress_input_3.txt
!./max_pairwise_sort_nlogn < stress_input_4.txt
!./max_pairwise_sort_nlogn < stress_input_5.txt
!./max_pairwise_sort_nlogn < stress_input_6.txt
!./max_pairwise_sort_nlogn < stress_input_7.txt
!./max_pairwise_sort_nlogn < stress_input_8.txt
!./max_pairwise_sort_nlogn < stress_input_9.txt
!./max_pairwise_sort_nlogn < stress_input_10.txt

!./max_pairwise_sort_nlogn < stress_input_11.txt
!./max_pairwise_sort_nlogn < stress_input_12.txt
!./max_pairwise_sort_nlogn < stress_input_13.txt
!./max_pairwise_sort_nlogn < stress_input_14.txt
!./max_pairwise_sort_nlogn < stress_input_15.txt
!./max_pairwise_sort_nlogn < stress_input_16.txt
!./max_pairwise_sort_nlogn < stress_input_17.txt
!./max_pairwise_sort_nlogn < stress_input_18.txt
!./max_pairwise_sort_nlogn < stress_input_19.txt
!./max_pairwise_sort_nlogn < stress_input_20.txt

!./max_pairwise_sort_nlogn < stress_input_21.txt
!./max_pairwise_sort_nlogn < stress_input_22.txt
!./max_pairwise_sort_nlogn < stress_input_23.txt
!./max_pairwise_sort_nlogn < stress_input_24.txt
!./max_pairwise_sort_nlogn < stress_input_25.txt
!./max_pairwise_sort_nlogn < stress_input_26.txt
!./max_pairwise_sort_nlogn < stress_input_27.txt
!./max_pairwise_sort_nlogn < stress_input_28.txt
!./max_pairwise_sort_nlogn < stress_input_29.txt
!./max_pairwise_sort_nlogn < stress_input_30.txt

!./max_pairwise_sort_nlogn < stress_input_31.txt
!./max_pairwise_sort_nlogn < stress_input_32.txt
!./max_pairwise_sort_nlogn < stress_input_33.txt
!./max_pairwise_sort_nlogn < stress_input_34.txt
!./max_pairwise_sort_nlogn < stress_input_35.txt
!./max_pairwise_sort_nlogn < stress_input_36.txt
!./max_pairwise_sort_nlogn < stress_input_37.txt
!./max_pairwise_sort_nlogn < stress_input_38.txt
!./max_pairwise_sort_nlogn < stress_input_39.txt
!./max_pairwise_sort_nlogn < stress_input_40.txt

!./max_pairwise_sort_nlogn < stress_input_41.txt
!./max_pairwise_sort_nlogn < stress_input_42.txt
!./max_pairwise_sort_nlogn < stress_input_43.txt
!./max_pairwise_sort_nlogn < stress_input_44.txt
!./max_pairwise_sort_nlogn < stress_input_45.txt
!./max_pairwise_sort_nlogn < stress_input_46.txt
!./max_pairwise_sort_nlogn < stress_input_47.txt
!./max_pairwise_sort_nlogn < stress_input_48.txt
!./max_pairwise_sort_nlogn < stress_input_49.txt
!./max_pairwise_sort_nlogn < stress_input_50.txt

!./max_pairwise_sort_nlogn < stress_input_51.txt
!./max_pairwise_sort_nlogn < stress_input_52.txt
!./max_pairwise_sort_nlogn < stress_input_53.txt
!./max_pairwise_sort_nlogn < stress_input_54.txt
!./max_pairwise_sort_nlogn < stress_input_55.txt
!./max_pairwise_sort_nlogn < stress_input_56.txt
!./max_pairwise_sort_nlogn < stress_input_57.txt
!./max_pairwise_sort_nlogn < stress_input_58.txt
!./max_pairwise_sort_nlogn < stress_input_59.txt
!./max_pairwise_sort_nlogn < stress_input_60.txt

!./max_pairwise_sort_nlogn < stress_input_61.txt
!./max_pairwise_sort_nlogn < stress_input_62.txt
!./max_pairwise_sort_nlogn < stress_input_63.txt
!./max_pairwise_sort_nlogn < stress_input_64.txt
!./max_pairwise_sort_nlogn < stress_input_65.txt
!./max_pairwise_sort_nlogn < stress_input_66.txt
!./max_pairwise_sort_nlogn < stress_input_67.txt
!./max_pairwise_sort_nlogn < stress_input_68.txt
!./max_pairwise_sort_nlogn < stress_input_69.txt
!./max_pairwise_sort_nlogn < stress_input_70.txt

!./max_pairwise_sort_nlogn < stress_input_71.txt
!./max_pairwise_sort_nlogn < stress_input_72.txt
!./max_pairwise_sort_nlogn < stress_input_73.txt
!./max_pairwise_sort_nlogn < stress_input_74.txt
!./max_pairwise_sort_nlogn < stress_input_75.txt
!./max_pairwise_sort_nlogn < stress_input_76.txt
!./max_pairwise_sort_nlogn < stress_input_77.txt
!./max_pairwise_sort_nlogn < stress_input_78.txt
!./max_pairwise_sort_nlogn < stress_input_79.txt
!./max_pairwise_sort_nlogn < stress_input_80.txt

!./max_pairwise_sort_nlogn < stress_input_81.txt
!./max_pairwise_sort_nlogn < stress_input_82.txt
!./max_pairwise_sort_nlogn < stress_input_83.txt
!./max_pairwise_sort_nlogn < stress_input_84.txt
!./max_pairwise_sort_nlogn < stress_input_85.txt
!./max_pairwise_sort_nlogn < stress_input_86.txt
!./max_pairwise_sort_nlogn < stress_input_87.txt
!./max_pairwise_sort_nlogn < stress_input_88.txt
!./max_pairwise_sort_nlogn < stress_input_89.txt
!./max_pairwise_sort_nlogn < stress_input_90.txt

!./max_pairwise_sort_nlogn < stress_input_91.txt
!./max_pairwise_sort_nlogn < stress_input_92.txt
!./max_pairwise_sort_nlogn < stress_input_93.txt
!./max_pairwise_sort_nlogn < stress_input_94.txt
!./max_pairwise_sort_nlogn < stress_input_95.txt
!./max_pairwise_sort_nlogn < stress_input_96.txt
!./max_pairwise_sort_nlogn < stress_input_97.txt
!./max_pairwise_sort_nlogn < stress_input_98.txt
!./max_pairwise_sort_nlogn < stress_input_99.txt
!./max_pairwise_sort_nlogn < stress_input_100.txt


n: 16 Product: 9025
n: 38 Product: 9024
n: 10 Product: 7728
n: 40 Product: 9300
n: 18 Product: 8366
n: 27 Product: 7920
n: 5 Product: 5160
n: 32 Product: 9312
n: 39 Product: 8835
n: 21 Product: 8740
n: 12 Product: 9215
n: 4 Product: 7990
n: 25 Product: 9800
n: 16 Product: 8190
n: 33 Product: 9409
n: 28 Product: 9504
n: 40 Product: 9603
n: 21 Product: 7031
n: 29 Product: 8099
n: 13 Product: 8051
n: 20 Product: 5508
n: 4 Product: 735
n: 22 Product: 9801
n: 29 Product: 9900
n: 36 Product: 9603
n: 4 Product: 1566
n: 36 Product: 9408
n: 25 Product: 9306
n: 32 Product: 8918
n: 9 Product: 7857
n: 17 Product: 8556
n: 34 Product: 10000
n: 19 Product: 7921
n: 3 Product: 2508
n: 29 Product: 8184
n: 8 Product: 8245
n: 14 Product: 8148
n: 17 Product: 7743
n: 10 Product: 8463
n: 12 Product: 7568
n: 10 Product: 4050
n: 38 Product: 9408
n: 12 Product: 6016
n: 13 Product: 8272
n: 8 Product: 7920
n: 28 Product: 8930
n: 32 Product: 9603
n: 26 Product: 8256
n: 30 Product: 8134
n: 9 Product: 5832
n: 13 Pro

# Python

## Super efficient solution (𝑇(𝑛)<10𝑛=𝑂(𝑛))

In [147]:
%%writefile max_pairwise_sort_super_efficient.py
n = int(input())                            # 1
max_number = float("-inf")                  # 1
second_max_number = float("-inf")           # 1

for num in input().split():                 # 3 + 2n

    num = int(num)                          # n

    if num > max_number:                    # n
        second_max_number = max_number      # [0, n] in the worst case it would be n times
        max_number = num                    # [0, n] in the worst case it would be n times
    elif num > second_max_number:           # n
        second_max_number = num             # [0, n] in the worst case it would be n times

print(f"n: {n} Product: {max_number * second_max_number}")

Overwriting max_pairwise_sort_super_efficient.py


In [148]:
%%time
!python max_pairwise_sort_super_efficient.py < stress_input_1.txt
!python max_pairwise_sort_super_efficient.py < stress_input_2.txt
!python max_pairwise_sort_super_efficient.py < stress_input_3.txt
!python max_pairwise_sort_super_efficient.py < stress_input_4.txt
!python max_pairwise_sort_super_efficient.py < stress_input_5.txt
!python max_pairwise_sort_super_efficient.py < stress_input_6.txt
!python max_pairwise_sort_super_efficient.py < stress_input_7.txt
!python max_pairwise_sort_super_efficient.py < stress_input_8.txt
!python max_pairwise_sort_super_efficient.py < stress_input_9.txt
!python max_pairwise_sort_super_efficient.py < stress_input_10.txt

!python max_pairwise_sort_super_efficient.py < stress_input_11.txt
!python max_pairwise_sort_super_efficient.py < stress_input_12.txt
!python max_pairwise_sort_super_efficient.py < stress_input_13.txt
!python max_pairwise_sort_super_efficient.py < stress_input_14.txt
!python max_pairwise_sort_super_efficient.py < stress_input_15.txt
!python max_pairwise_sort_super_efficient.py < stress_input_16.txt
!python max_pairwise_sort_super_efficient.py < stress_input_17.txt
!python max_pairwise_sort_super_efficient.py < stress_input_18.txt
!python max_pairwise_sort_super_efficient.py < stress_input_19.txt
!python max_pairwise_sort_super_efficient.py < stress_input_20.txt

!python max_pairwise_sort_super_efficient.py < stress_input_21.txt
!python max_pairwise_sort_super_efficient.py < stress_input_22.txt
!python max_pairwise_sort_super_efficient.py < stress_input_23.txt
!python max_pairwise_sort_super_efficient.py < stress_input_24.txt
!python max_pairwise_sort_super_efficient.py < stress_input_25.txt
!python max_pairwise_sort_super_efficient.py < stress_input_26.txt
!python max_pairwise_sort_super_efficient.py < stress_input_27.txt
!python max_pairwise_sort_super_efficient.py < stress_input_28.txt
!python max_pairwise_sort_super_efficient.py < stress_input_29.txt
!python max_pairwise_sort_super_efficient.py < stress_input_30.txt

!python max_pairwise_sort_super_efficient.py < stress_input_31.txt
!python max_pairwise_sort_super_efficient.py < stress_input_32.txt
!python max_pairwise_sort_super_efficient.py < stress_input_33.txt
!python max_pairwise_sort_super_efficient.py < stress_input_34.txt
!python max_pairwise_sort_super_efficient.py < stress_input_35.txt
!python max_pairwise_sort_super_efficient.py < stress_input_36.txt
!python max_pairwise_sort_super_efficient.py < stress_input_37.txt
!python max_pairwise_sort_super_efficient.py < stress_input_38.txt
!python max_pairwise_sort_super_efficient.py < stress_input_39.txt
!python max_pairwise_sort_super_efficient.py < stress_input_40.txt

!python max_pairwise_sort_super_efficient.py < stress_input_41.txt
!python max_pairwise_sort_super_efficient.py < stress_input_42.txt
!python max_pairwise_sort_super_efficient.py < stress_input_43.txt
!python max_pairwise_sort_super_efficient.py < stress_input_44.txt
!python max_pairwise_sort_super_efficient.py < stress_input_45.txt
!python max_pairwise_sort_super_efficient.py < stress_input_46.txt
!python max_pairwise_sort_super_efficient.py < stress_input_47.txt
!python max_pairwise_sort_super_efficient.py < stress_input_48.txt
!python max_pairwise_sort_super_efficient.py < stress_input_49.txt
!python max_pairwise_sort_super_efficient.py < stress_input_50.txt

!python max_pairwise_sort_super_efficient.py < stress_input_51.txt
!python max_pairwise_sort_super_efficient.py < stress_input_52.txt
!python max_pairwise_sort_super_efficient.py < stress_input_53.txt
!python max_pairwise_sort_super_efficient.py < stress_input_54.txt
!python max_pairwise_sort_super_efficient.py < stress_input_55.txt
!python max_pairwise_sort_super_efficient.py < stress_input_56.txt
!python max_pairwise_sort_super_efficient.py < stress_input_57.txt
!python max_pairwise_sort_super_efficient.py < stress_input_58.txt
!python max_pairwise_sort_super_efficient.py < stress_input_59.txt
!python max_pairwise_sort_super_efficient.py < stress_input_60.txt

!python max_pairwise_sort_super_efficient.py < stress_input_61.txt
!python max_pairwise_sort_super_efficient.py < stress_input_62.txt
!python max_pairwise_sort_super_efficient.py < stress_input_63.txt
!python max_pairwise_sort_super_efficient.py < stress_input_64.txt
!python max_pairwise_sort_super_efficient.py < stress_input_65.txt
!python max_pairwise_sort_super_efficient.py < stress_input_66.txt
!python max_pairwise_sort_super_efficient.py < stress_input_67.txt
!python max_pairwise_sort_super_efficient.py < stress_input_68.txt
!python max_pairwise_sort_super_efficient.py < stress_input_69.txt
!python max_pairwise_sort_super_efficient.py < stress_input_70.txt

!python max_pairwise_sort_super_efficient.py < stress_input_71.txt
!python max_pairwise_sort_super_efficient.py < stress_input_72.txt
!python max_pairwise_sort_super_efficient.py < stress_input_73.txt
!python max_pairwise_sort_super_efficient.py < stress_input_74.txt
!python max_pairwise_sort_super_efficient.py < stress_input_75.txt
!python max_pairwise_sort_super_efficient.py < stress_input_76.txt
!python max_pairwise_sort_super_efficient.py < stress_input_77.txt
!python max_pairwise_sort_super_efficient.py < stress_input_78.txt
!python max_pairwise_sort_super_efficient.py < stress_input_79.txt
!python max_pairwise_sort_super_efficient.py < stress_input_80.txt

!python max_pairwise_sort_super_efficient.py < stress_input_81.txt
!python max_pairwise_sort_super_efficient.py < stress_input_82.txt
!python max_pairwise_sort_super_efficient.py < stress_input_83.txt
!python max_pairwise_sort_super_efficient.py < stress_input_84.txt
!python max_pairwise_sort_super_efficient.py < stress_input_85.txt
!python max_pairwise_sort_super_efficient.py < stress_input_86.txt
!python max_pairwise_sort_super_efficient.py < stress_input_87.txt
!python max_pairwise_sort_super_efficient.py < stress_input_88.txt
!python max_pairwise_sort_super_efficient.py < stress_input_89.txt
!python max_pairwise_sort_super_efficient.py < stress_input_90.txt

!python max_pairwise_sort_super_efficient.py < stress_input_91.txt
!python max_pairwise_sort_super_efficient.py < stress_input_92.txt
!python max_pairwise_sort_super_efficient.py < stress_input_93.txt
!python max_pairwise_sort_super_efficient.py < stress_input_94.txt
!python max_pairwise_sort_super_efficient.py < stress_input_95.txt
!python max_pairwise_sort_super_efficient.py < stress_input_96.txt
!python max_pairwise_sort_super_efficient.py < stress_input_97.txt
!python max_pairwise_sort_super_efficient.py < stress_input_98.txt
!python max_pairwise_sort_super_efficient.py < stress_input_99.txt
!python max_pairwise_sort_super_efficient.py < stress_input_100.txt


n: 16 Product: 9025
n: 38 Product: 9024
n: 10 Product: 7728
n: 40 Product: 9300
n: 18 Product: 8366
n: 27 Product: 7920
n: 5 Product: 5160
n: 32 Product: 9312
n: 39 Product: 8835
n: 21 Product: 8740
n: 12 Product: 9215
n: 4 Product: 7990
n: 25 Product: 9800
n: 16 Product: 8190
n: 33 Product: 9409
n: 28 Product: 9504
n: 40 Product: 9603
n: 21 Product: 7031
n: 29 Product: 8099
n: 13 Product: 8051
n: 20 Product: 5508
n: 4 Product: 735
n: 22 Product: 9801
n: 29 Product: 9900
n: 36 Product: 9603
n: 4 Product: 1566
n: 36 Product: 9408
n: 25 Product: 9306
n: 32 Product: 8918
n: 9 Product: 7857
n: 17 Product: 8556
n: 34 Product: 10000
n: 19 Product: 7921
n: 3 Product: 2508
n: 29 Product: 8184
n: 8 Product: 8245
n: 14 Product: 8148
n: 17 Product: 7743
n: 10 Product: 8463
n: 12 Product: 7568
n: 10 Product: 4050
n: 38 Product: 9408
n: 12 Product: 6016
n: 13 Product: 8272
n: 8 Product: 7920
n: 28 Product: 8930
n: 32 Product: 9603
n: 26 Product: 8256
n: 30 Product: 8134
n: 9 Product: 5832
n: 13 Pro

## Sort solution (𝑇(𝑛)=𝑂(𝑛𝑙𝑛𝑔))

In [149]:
%%writefile max_pairwise_sort_nlogn.py
# Read input values
n = int(input())              # 1
numbers = list(map(int, input().split()))  # n

# Sort the list
numbers.sort()                # n*log(n)

# Calculate the maximum pairwise product
result = numbers[-1] * numbers[-2]  # 1

# Print the result
print(f"n: {n} Product: {result}")

Overwriting max_pairwise_sort_nlogn.py


In [150]:
%%time
!python max_pairwise_sort_nlogn.py < stress_input_1.txt
!python max_pairwise_sort_nlogn.py < stress_input_2.txt
!python max_pairwise_sort_nlogn.py < stress_input_3.txt
!python max_pairwise_sort_nlogn.py < stress_input_4.txt
!python max_pairwise_sort_nlogn.py < stress_input_5.txt
!python max_pairwise_sort_nlogn.py < stress_input_6.txt
!python max_pairwise_sort_nlogn.py < stress_input_7.txt
!python max_pairwise_sort_nlogn.py < stress_input_8.txt
!python max_pairwise_sort_nlogn.py < stress_input_9.txt
!python max_pairwise_sort_nlogn.py < stress_input_10.txt

!python max_pairwise_sort_nlogn.py < stress_input_11.txt
!python max_pairwise_sort_nlogn.py < stress_input_12.txt
!python max_pairwise_sort_nlogn.py < stress_input_13.txt
!python max_pairwise_sort_nlogn.py < stress_input_14.txt
!python max_pairwise_sort_nlogn.py < stress_input_15.txt
!python max_pairwise_sort_nlogn.py < stress_input_16.txt
!python max_pairwise_sort_nlogn.py < stress_input_17.txt
!python max_pairwise_sort_nlogn.py < stress_input_18.txt
!python max_pairwise_sort_nlogn.py < stress_input_19.txt
!python max_pairwise_sort_nlogn.py < stress_input_20.txt

!python max_pairwise_sort_nlogn.py < stress_input_21.txt
!python max_pairwise_sort_nlogn.py < stress_input_22.txt
!python max_pairwise_sort_nlogn.py < stress_input_23.txt
!python max_pairwise_sort_nlogn.py < stress_input_24.txt
!python max_pairwise_sort_nlogn.py < stress_input_25.txt
!python max_pairwise_sort_nlogn.py < stress_input_26.txt
!python max_pairwise_sort_nlogn.py < stress_input_27.txt
!python max_pairwise_sort_nlogn.py < stress_input_28.txt
!python max_pairwise_sort_nlogn.py < stress_input_29.txt
!python max_pairwise_sort_nlogn.py < stress_input_30.txt

!python max_pairwise_sort_nlogn.py < stress_input_31.txt
!python max_pairwise_sort_nlogn.py < stress_input_32.txt
!python max_pairwise_sort_nlogn.py < stress_input_33.txt
!python max_pairwise_sort_nlogn.py < stress_input_34.txt
!python max_pairwise_sort_nlogn.py < stress_input_35.txt
!python max_pairwise_sort_nlogn.py < stress_input_36.txt
!python max_pairwise_sort_nlogn.py < stress_input_37.txt
!python max_pairwise_sort_nlogn.py < stress_input_38.txt
!python max_pairwise_sort_nlogn.py < stress_input_39.txt
!python max_pairwise_sort_nlogn.py < stress_input_40.txt

!python max_pairwise_sort_nlogn.py < stress_input_41.txt
!python max_pairwise_sort_nlogn.py < stress_input_42.txt
!python max_pairwise_sort_nlogn.py < stress_input_43.txt
!python max_pairwise_sort_nlogn.py < stress_input_44.txt
!python max_pairwise_sort_nlogn.py < stress_input_45.txt
!python max_pairwise_sort_nlogn.py < stress_input_46.txt
!python max_pairwise_sort_nlogn.py < stress_input_47.txt
!python max_pairwise_sort_nlogn.py < stress_input_48.txt
!python max_pairwise_sort_nlogn.py < stress_input_49.txt
!python max_pairwise_sort_nlogn.py < stress_input_50.txt

!python max_pairwise_sort_nlogn.py < stress_input_51.txt
!python max_pairwise_sort_nlogn.py < stress_input_52.txt
!python max_pairwise_sort_nlogn.py < stress_input_53.txt
!python max_pairwise_sort_nlogn.py < stress_input_54.txt
!python max_pairwise_sort_nlogn.py < stress_input_55.txt
!python max_pairwise_sort_nlogn.py < stress_input_56.txt
!python max_pairwise_sort_nlogn.py < stress_input_57.txt
!python max_pairwise_sort_nlogn.py < stress_input_58.txt
!python max_pairwise_sort_nlogn.py < stress_input_59.txt
!python max_pairwise_sort_nlogn.py < stress_input_60.txt

!python max_pairwise_sort_nlogn.py < stress_input_61.txt
!python max_pairwise_sort_nlogn.py < stress_input_62.txt
!python max_pairwise_sort_nlogn.py < stress_input_63.txt
!python max_pairwise_sort_nlogn.py < stress_input_64.txt
!python max_pairwise_sort_nlogn.py < stress_input_65.txt
!python max_pairwise_sort_nlogn.py < stress_input_66.txt
!python max_pairwise_sort_nlogn.py < stress_input_67.txt
!python max_pairwise_sort_nlogn.py < stress_input_68.txt
!python max_pairwise_sort_nlogn.py < stress_input_69.txt
!python max_pairwise_sort_nlogn.py < stress_input_70.txt

!python max_pairwise_sort_nlogn.py < stress_input_71.txt
!python max_pairwise_sort_nlogn.py < stress_input_72.txt
!python max_pairwise_sort_nlogn.py < stress_input_73.txt
!python max_pairwise_sort_nlogn.py < stress_input_74.txt
!python max_pairwise_sort_nlogn.py < stress_input_75.txt
!python max_pairwise_sort_nlogn.py < stress_input_76.txt
!python max_pairwise_sort_nlogn.py < stress_input_77.txt
!python max_pairwise_sort_nlogn.py < stress_input_78.txt
!python max_pairwise_sort_nlogn.py < stress_input_79.txt
!python max_pairwise_sort_nlogn.py < stress_input_80.txt

!python max_pairwise_sort_nlogn.py < stress_input_81.txt
!python max_pairwise_sort_nlogn.py < stress_input_82.txt
!python max_pairwise_sort_nlogn.py < stress_input_83.txt
!python max_pairwise_sort_nlogn.py < stress_input_84.txt
!python max_pairwise_sort_nlogn.py < stress_input_85.txt
!python max_pairwise_sort_nlogn.py < stress_input_86.txt
!python max_pairwise_sort_nlogn.py < stress_input_87.txt
!python max_pairwise_sort_nlogn.py < stress_input_88.txt
!python max_pairwise_sort_nlogn.py < stress_input_89.txt
!python max_pairwise_sort_nlogn.py < stress_input_90.txt

!python max_pairwise_sort_nlogn.py < stress_input_91.txt
!python max_pairwise_sort_nlogn.py < stress_input_92.txt
!python max_pairwise_sort_nlogn.py < stress_input_93.txt
!python max_pairwise_sort_nlogn.py < stress_input_94.txt
!python max_pairwise_sort_nlogn.py < stress_input_95.txt
!python max_pairwise_sort_nlogn.py < stress_input_96.txt
!python max_pairwise_sort_nlogn.py < stress_input_97.txt
!python max_pairwise_sort_nlogn.py < stress_input_98.txt
!python max_pairwise_sort_nlogn.py < stress_input_99.txt
!python max_pairwise_sort_nlogn.py < stress_input_100.txt


n: 16 Product: 9025
n: 38 Product: 9024
n: 10 Product: 7728
n: 40 Product: 9300
n: 18 Product: 8366
n: 27 Product: 7920
n: 5 Product: 5160
n: 32 Product: 9312
n: 39 Product: 8835
n: 21 Product: 8740
n: 12 Product: 9215
n: 4 Product: 7990
n: 25 Product: 9800
n: 16 Product: 8190
n: 33 Product: 9409
n: 28 Product: 9504
n: 40 Product: 9603
n: 21 Product: 7031
n: 29 Product: 8099
n: 13 Product: 8051
n: 20 Product: 5508
n: 4 Product: 735
n: 22 Product: 9801
n: 29 Product: 9900
n: 36 Product: 9603
n: 4 Product: 1566
n: 36 Product: 9408
n: 25 Product: 9306
n: 32 Product: 8918
n: 9 Product: 7857
n: 17 Product: 8556
n: 34 Product: 10000
n: 19 Product: 7921
n: 3 Product: 2508
n: 29 Product: 8184
n: 8 Product: 8245
n: 14 Product: 8148
n: 17 Product: 7743
n: 10 Product: 8463
n: 12 Product: 7568
n: 10 Product: 4050
n: 38 Product: 9408
n: 12 Product: 6016
n: 13 Product: 8272
n: 8 Product: 7920
n: 28 Product: 8930
n: 32 Product: 9603
n: 26 Product: 8256
n: 30 Product: 8134
n: 9 Product: 5832
n: 13 Pro

# Resultados

In [151]:
# Import the Pandas library
import pandas as pd

# Create a sample DataFrame
data = {'Language': ['C++', 'Python'],
        'Super effienct (ms)': [539, 516],
        'Sort solution (ms)': [542, 485]}

df = pd.DataFrame(data)

# Display the table
display(df)

Language  Super effienct (ms)  Sort solution (ms)
0      C++                  539                 542
1   Python                  516                 485